drivers/sensor/vcnl4040/vcnl4040.c

```diff
@@ -0,0 +1,358 @@
+
+int vcnl4040_read(struct device *dev, uint8_t reg, uint16_t *out)
+{
+	const struct vcnl4040_config *config = dev->config;
+	struct vcnl4040_data *data = dev->data;
+
+	uint8_t buff[2] = { 0 };
+	int ret = 0;
+
+	ret = i2c_write_read(data->i2c, config->i2c_address,
+			     &reg, sizeof(reg), buff, sizeof(buff));
+
+	*out = sys_get_le16(buff);
+
+	return ret;
+}
+
+int vcnl4040_write(struct device *dev, uint8_t reg, uint16_t value)
+{
+	const struct vcnl4040_config *config = dev->config;
+	struct vcnl4040_data *data = dev->data;
+
+	struct i2c_msg msg;
+	int ret;
+	uint8_t buff[3];
+
+	sys_put_le16(value, &buff[1]);
+
+	buff[0] = reg;
+
+	msg.buf = buff;
+	msg.flags = 0;
+	msg.len = sizeof(buff);
+
+	ret = i2c_transfer(data->i2c, &msg, 1, config->i2c_address);
+
+	if (ret < 0) {
+		LOG_ERR("write block failed");
+		return -1;
+	}
+
+	return 0;
+}
+
+static int vcnl4040_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct vcnl4040_data *data = dev->data;
+	int ret = 0;
+
+	if (chan != SENSOR_CHAN_ALL) {
+		LOG_ERR("Unsupported sensor channel");
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_DATA, &data->proximity)) {
+		LOG_ERR("Could not fetch proximity");
+		ret = -EIO;
+		goto exit;
+	}
+#ifdef CONFIG_VCNL4040_ENABLE_ALS
+	const struct vcnl4040_config *config = dev->config;
+
+	if (vcnl4040_read(dev, VCNL4040_REG_ALS_DATA, &data->light)) {
+		LOG_ERR("Could not fetch ambient light");
+		ret = -EIO;
+		goto exit;
+	}
+	/*
+	 * scale the lux depending on the value of the integration time
+	 * see page 8 of the VCNL4040 application note:
+	 * https://www.vishay.com/docs/84307/designingvcnl4040.pdf
+	 */
+	data->light = (data->light * (0.1 / (1 << config->als_it)));
+#endif
+exit:
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int vcnl4040_channel_get(struct device *dev,
+				enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	struct vcnl4040_data *data = dev->data;
+	int ret = 0;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	switch (chan) {
+	case SENSOR_CHAN_PROX:
+		val->val1 = data->proximity;
+		val->val2 = 0;
+		break;
+
+#ifdef CONFIG_VCNL4040_ENABLE_ALS
+	case SENSOR_CHAN_LIGHT:
+		val->val1 = data->light;
+		val->val2 = 0;
+		break;
+#endif
+
+	default:
+		ret = -ENOTSUP;
+	}
+
+	k_sem_give(&data->sem);
+
+	return ret;
+}
+
+static int vcnl4040_proxy_setup(struct device *dev)
+{
+	const struct vcnl4040_config *config = dev->config;
+	uint16_t conf = 0;
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_MS, &conf)) {
+		LOG_ERR("Could not read proximity config");
+		return -EIO;
+	}
+
+	/* Set LED current bits 2:0 of high register */
+	conf |= config->led_i << 8;
+
+	if (vcnl4040_write(dev, VCNL4040_REG_PS_MS, conf)) {
+		LOG_ERR("Could not write proximity config");
+		return -EIO;
+	}
+
+	if (vcnl4040_read(dev, VCNL4040_REG_PS_CONF, &conf)) {
+		LOG_ERR("Could not read proximity config");
+		return -EIO;
+	}
+
+	/* Set PS_HD bit 3 of high register */
```

Comment:
Since you define a driver-specific header the magic numbers here should go into it.  The common idiom is to define `_POS` and `_MASK` values, e.g. https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/can/can_mcp2515.h#L157

Reply:
Added POS and MASK as requested